streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1486,6 +1726,35 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
   final def transduce[R1 <: R, E1 >: E, A1 >: A, C](sink: ZSink[R1, E1, A1, A1, C]): ZStream[R1, E1, C] =
     transduceManaged[R1, E1, A1, C](ZManaged.succeed(sink))
 
+  /**
+   * Acquire a permit for every consumer task.
+   */
+  final def withPermit(sema: Semaphore): ZStream[R, E, A] =
+    withPermits(sema, 1)
+
+  /**
+   * Acquire n permits for every consumer task.
+   */
+  final def withPermits(sema: Semaphore, n: Long): ZStream[R, E, A] =
+    new ZStream[R, E, A] {
+      def fold[R1 <: R, E1 >: E, A1 >: A, S]: ZStream.Fold[R1, E1, A1, S] =
+        ZManaged.effectTotal { (s, cont, f) =>
+          self.fold[R1, E1, A1, S].flatMap { fold =>
+            fold(s, cont, (s, a) => sema.withPermits(n)(f(s, a)))
+          }
+        }
+    }
+
+  /**
+   * Adds a shared semaphore with n permits to the stream
+   */
+  final def withSemaphore(n: Long): ZStream[R, E, (Semaphore, A)] =
```

Comment:
Not sold on this. This is can be just `ZStream.fromEffect(Semaphore.make(n)).flatMap { sem =>`

Reply:
Agreed